File: packages/cli/jack/cli.kt/src/main/kotlin/io/github/hieudoanm/cli/commands/Doi.kt
package io.github.hieudoanm.cli.commands

import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.subcommands
import com.github.ajalt.clikt.parameters.arguments.argument
import com.google.gson.GsonBuilder
import com.google.gson.JsonParser
import io.github.hieudoanm.cli.services.Requests

class DoiCommand : CliktCommand(name = "doi", help = "DOI productivity tools") {
    init {
        subcommands(DoiCite(), DoiFetch(), DoiRef(), DoiValidate())
    }
    override fun run() = Unit
}

class DoiCite : CliktCommand(name = "cite", help = "Generate an APA citation from a DOI") {
    private val doi by argument()
    override fun run() {
        val response = fetchCrossref(doi)
        if (response == null) return
        echo(citationText(response))
    }
}

class DoiFetch : CliktCommand(name = "fetch", help = "Fetch raw metadata for a DOI") {
    private val doi by argument()
    override fun run() {
        val response = fetchCrossref(doi)
        if (response != null) {
            val g = GsonBuilder().setPrettyPrinting().create()
            echo(g.toJson(JsonParser.parseString(response)))
        }
    }
}

class DoiRef : CliktCommand(name = "ref", help = "Generate a formatted reference from a DOI") {
    private val doi by argument()
    override fun run() {
        val response = fetchCrossref(doi)
        if (response == null) return
        echo(referenceText(response))
    }
}

class DoiValidate : CliktCommand(name = "validate", help = "Validate a DOI string format") {
    private val doi by argument()
    override fun run() {
        val matched = Regex("^10\\.\\d{4,}/.+$").matches(doi)
        if (matched) {
            echo("\u2713 $doi is a valid DOI")
        } else {
            echo("\u2717 $doi is not a valid DOI format")
        }
    }
}

private fun fetchCrossref(doi: String): String? {
    val url = "https://api.crossref.org/works/$doi"
    val result = Requests.get(url)
    if (result.isFailure) {
        return null
    }
    return result.getOrNull()
}

private fun citationText(response: String): String {
    val root = JsonParser.parseString(response).asJsonObject
    val message = root.getAsJsonObject("message")

    val authors = message.getAsJsonArray("author")?.map { it.asJsonObject } ?: emptyList()
    val year = try {
        message.getAsJsonObject("published-print")
            ?.getAsJsonArray("date-parts")
            ?.get(0)?.asJsonArray?.get(0)?.asInt?.toString() ?: "n.d."
    } catch (e: Exception) { "n.d." }

    val citation = when {
        authors.isEmpty() -> "(Unknown, $year)"
        authors.size == 1 -> "(${authors[0].get("family")?.asString}, $year)"
        authors.size == 2 -> "(${authors[0].get("family")?.asString} & ${authors[1].get("family")?.asString}, $year)"
        else -> "(${authors[0].get("family")?.asString} et al., $year)"
    }
    return "Cite:\n$citation"
}

private fun referenceText(response: String): String {
    val root = JsonParser.parseString(response).asJsonObject
    val message = root.getAsJsonObject("message")
    val authors = message.getAsJsonArray("author")?.map { it.asJsonObject } ?: emptyList()

    val authorsAPA = authors.mapIndexed { i, a ->
        val given = a.get("given")?.asString ?: ""
        val family = a.get("family")?.asString ?: ""
        val initial = if (given.isNotEmpty()) "${given.first()}." else ""
        when {
            i == 0 -> "$family, $initial"
            i == authors.size - 1 -> " & $family, $initial"
            else -> ", $family, $initial"
        }
    }.joinToString("")

    val year = try {
        message.getAsJsonObject("published-print")
            ?.getAsJsonArray("date-parts")
            ?.get(0)?.asJsonArray?.get(0)?.asInt?.toString() ?: "n.d."
    } catch (e: Exception) { "n.d." }

    val title = message.getAsJsonArray("title")?.get(0)?.asString ?: ""
    val journal = message.getAsJsonArray("container-title")?.get(0)?.asString ?: ""
    val volume = message.get("volume")?.asString ?: ""
    val issue = message.get("issue")?.asString ?: ""
    val pages = message.get("page")?.asString ?: ""

    val apa = "$authorsAPA ($year). $title. $journal, $volume($issue), $pages."
    return "APA:\n$apa"
}

File: packages/app/headless/jack/languages/swift/Sources/jack/Commands/Doi/Doi.swift
import Foundation
import ArgumentParser

// MARK: - Crossref API Types

struct CrossrefResponse: Codable {
    let message: CrossrefMessage?
}

struct CrossrefMessage: Codable {
    let author: [CrossrefAuthor]?
    let title: [String]?
    let containerTitle: [String]?
    let volume: String?
    let issue: String?
    let page: String?
    let publishedOnline: CrossrefDate?
    let publishedPrint: CrossrefDate?
    let publisher: String?
    let type: String?

    enum CodingKeys: String, CodingKey {
        case author, title, volume, issue, page, publisher, type
        case containerTitle = "container-title"
        case publishedOnline = "published-online"
        case publishedPrint = "published-print"
    }
}

struct CrossrefAuthor: Codable {
    let given: String?
    let family: String?
}

struct CrossrefDate: Codable {
    let dateParts: [[Int]]?

    enum CodingKeys: String, CodingKey {
        case dateParts = "date-parts"
    }
}

private func fetchDoiMetadata(doi: String) async throws -> CrossrefMessage {
    let response: CrossrefResponse = try await Requests.fetchJSON(
        "https://api.crossref.org/works/\(doi)",
        headers: ["Accept": "application/json"]
    )
    guard let message = response.message else {
        throw RequestError(message: "No data found for DOI: \(doi)", statusCode: nil)
    }
    return message
}

func formatAuthors(_ authors: [CrossrefAuthor]?) -> String {
    guard let authors = authors, !authors.isEmpty else { return "Unknown" }
    if authors.count == 1 {
        let a = authors[0]
        return "\(a.family ?? "?"), \(String((a.given ?? "?").prefix(1)))."
    } else if authors.count == 2 {
        let a1 = authors[0], a2 = authors[1]
        return "\(a1.family ?? "?") & \(a2.family ?? "?")"
    } else {
        let first = authors[0]
        return "\(first.family ?? "?") et al."
    }
}

func formatAuthorsFull(_ authors: [CrossrefAuthor]?) -> String {
    guard let authors = authors, !authors.isEmpty else { return "Unknown" }
    return authors.map { "\($0.family ?? "?"), \(String(($0.given ?? "?").prefix(1)))." }.joined(separator: ", ")
}

func getYear(_ msg: CrossrefMessage) -> String {
    if let date = msg.publishedOnline ?? msg.publishedPrint,
       let parts = date.dateParts, let first = parts.first, let year = first.first {
        return "\(year)"
    }
    return "n.d."
}

struct DoiCommand: ParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "doi",
        abstract: "DOI citation tools",
        subcommands: [CiteCommand.self, RefCommand.self, FetchCommand.self, ValidateCommand.self]
    )
    mutating func run() {}
}

// MARK: - Cite

struct CiteCommand: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "cite", abstract: "Generate APA in-text citation from DOI")

    @Argument(help: "DOI identifier")
    var doi: String

    mutating func run() async throws {
        let msg = try await fetchDoiMetadata(doi: doi)
        let authors = formatAuthors(msg.author)
        let year = getYear(msg)
        print("(\(authors), \(year))")
    }
}

// MARK: - Ref

struct RefCommand: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "ref", abstract: "Generate APA reference entry from DOI")

    @Argument(help: "DOI identifier")
    var doi: String

    mutating func run() async throws {
        let msg = try await fetchDoiMetadata(doi: doi)
        let authors = formatAuthorsFull(msg.author)
        let year = getYear(msg)
        let title = msg.title?.first ?? "Untitled"
        let container = msg.containerTitle?.first ?? ""
        let publisher = msg.publisher ?? ""
        let volume = msg.volume != nil ? ", \(msg.volume!)" : ""
        let issue = msg.issue != nil ? "(\(msg.issue!))" : ""
        let pages = msg.page != nil ? ", \(msg.page!)" : ""

        var ref = "\(authors) (\(year)). \(title)."
        if !container.isEmpty {
            ref += " *\(container)*\(volume)\(issue)\(pages)."
        }
        ref += " \(publisher)."
        ref += " https://doi.org/\(doi)"
        print(ref)
    }
}

// MARK: - Fetch

struct FetchCommand: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "fetch", abstract: "Fetch DOI metadata from Crossref API")

    @Argument(help: "DOI identifier")
    var doi: String

    mutating func run() async throws {
        let url = "https://api.crossref.org/works/\(doi)"
        let response = try await Requests.fetch(url)
        let json = try JSONSerialization.jsonObject(with: response.data)
        let pretty = try JSONSerialization.data(withJSONObject: json, options: [.prettyPrinted])
        print(String(data: pretty, encoding: .utf8) ?? "")
    }
}

// MARK: - Validate

struct ValidateCommand: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "validate", abstract: "Validate DOI format and resolve")

    @Argument(help: "DOI identifier")
    var doi: String

    mutating func run() async throws {
        guard !doi.isEmpty else {
            print("invalid: empty DOI")
            return
        }

        let doiPattern = try NSRegularExpression(pattern: "^10\\.\\d{4,}/\\S+$")
        let range = NSRange(doi.startIndex..., in: doi)
        guard doiPattern.firstMatch(in: doi, range: range) != nil else {
            print("invalid: DOI must match pattern 10.xxxx/xxxxx")
            return
        }

        let url = "https://doi.org/\(doi)"
        do {
            let response = try await Requests.fetch(url)
            if response.statusCode == 200 || response.statusCode == 302 {
                print("valid: resolves to \(url)")
            } else {
                print("invalid: HTTP \(response.statusCode)")
            }
        } catch {
            print("invalid: \(error.localizedDescription)")
        }
    }
}
